pkg/danmep/interface.go

```diff
@@ -0,0 +1,73 @@
+// when deleting dangling DanmEps
+func GetSupportedReleaseIPHandlers() []*ReleaseIPService {
+    return []*ReleaseIPService{
+        NewReleaseIPService(&calicoReleaseIPServiceImpl{}),
+    }
+}
+
+// ReleaseIPInterface need to be implemented by every static CNI IPAM release plugin
+// that want to invoked when Cleaner is cleaning up dangling DanmEps
+type ReleaseIPInterface interface {
+    ReleaseIP(dnet *danmtypes.DanmNet, ep *danmtypes.DanmEp, ip string) error
+    IsIPAllocatedByMe(dnet *danmtypes.DanmNet, ep *danmtypes.DanmEp, ip string) bool
+}
+
+type ReleaseIPService struct {
+    handler ReleaseIPInterface
+}
+
+// NewReleaseIPService constructor encapsulates different ReleaseIPInterface implementations under a generic API
+func NewReleaseIPService(handler ReleaseIPInterface) *ReleaseIPService {
+    return &ReleaseIPService{handler}
+}
+
+// IsIPAllocatedByMe invokes embedded implementation specific IsIPAllocatedByMe function
+func (d ReleaseIPService) IsIPAllocatedByMe(dnet *danmtypes.DanmNet, ep *danmtypes.DanmEp, ip string) bool {
+    return ! ipam.WasIpAllocatedByDanm(ip, dnet.Spec.Options.Cidr) && d.handler.IsIPAllocatedByMe(dnet, ep, ip)
+}
+
+// ReleaseIP invokes embedded implementation specific DeleteDanmEp function
+func (d ReleaseIPService) ReleaseIP(dnet *danmtypes.DanmNet, ep *danmtypes.DanmEp, ip string) error {
+    return d.handler.ReleaseIP( dnet, ep, ip)
+}
+
+// SelectAppropriateServiceImplementation looks up and returns the
+// first suitable ReleaseIP service implementation for given IP address
+// returns nil when no suitable implementation found
+func SelectAppropriateServiceImplementation(dnet *danmtypes.DanmNet, ep *danmtypes.DanmEp, ip string) *ReleaseIPService {
+    for _, handler := range GetSupportedReleaseIPHandlers() {
+        if handler.IsIPAllocatedByMe(dnet, ep, ip) {
+            return handler
+        }
+    }
+    return nil
+}
+
+// DeleteDanmEp selects an ReleaseIPService implementation for the allocated IPv4 and IPv6 IP addresses
+// in a given DanmEp and invokes them separately trying ot free up allocated IPs, after successfully freeing up
+// the IPs, it deletes the DanmEp or returns error if unable to free any of the IPs or unable todelete DanmEp
+func DeleteDanmEp(danmClient danmclientset.Interface, ep *danmtypes.DanmEp, dnet *danmtypes.DanmNet) error {
+    if service := SelectAppropriateServiceImplementation(dnet, ep, ep.Spec.Iface.Address); service != nil {
+        if err := service.ReleaseIP(dnet, ep, ep.Spec.Iface.Address); err != nil {
+            return fmt.Errorf("unable to release ipv4 IP because: %s", err)
+        }
+    }
+    if service := SelectAppropriateServiceImplementation(dnet, ep, ep.Spec.Iface.AddressIPv6); service != nil {
+        if err := service.ReleaseIP(dnet, ep, ep.Spec.Iface.AddressIPv6); err != nil {
+            return fmt.Errorf("unable to release ipv6 IP because: %s", err)
+        }
+    }
+    return danmep.DeleteDanmEp(danmClient, ep, dnet)
```

Comment:
you really only need ths one line: return danmClient.DanmV1().DanmEps(ep.ObjectMeta.Namespace).Delete(context.TODO(), ep.ObjectMeta.Name, meta_v1.DeleteOptions{})
unnecessairy to again invoke the whole function which might even throw errors sometimes cause you are triying to clean up things twice

Reply:
or you again went back to not handling DANM IPAM as a normal handler and this is the actual DANM IPAM cleaning call too?